src/phasorpy/phasor.py

```
@@ -2998,6 +3000,168 @@ def phasor_filter_median(
+    normalize : bool, optional
+        By default (False), expects and returns normalized phasor coordinates.
+        If True, unnormalized phasor coordinates are expected as input.
+    samples : int, optional
+        Number of samples used to calculate the phasor coordinates.
+        By default, 1.
+    nan_safe : bool, optional
+        By default, distribute NaNs among filtered arrays.
+
+    Returns
+    -------
+    mean : ndarray
+        Unchanged intensity of phasor coordinates.
+    real : ndarray
+        Filtered real component of phasor coordinates.
+    imag : ndarray
+        Filtered imaginary component of phasor coordinates.
+
+    Raises
+    ------
+    ValueError
+        If `level` is less than 0.
+        If `samples` is less than 1.
+        If the number of harmonics in `harmonic` is less than 2.
+        If `real` and `imag` have no harmonic axis.
+        The array shapes of `mean`, `real`, and `imag` do not match.
+
+    Examples
+    --------
+
+    """
+    from pawflim import pawflim  # type: ignore[import-untyped]
+
+    mean = numpy.asarray(mean)
+    real = numpy.asarray(real)
+    imag = numpy.asarray(imag)
+
+    if levels < 0:
+        raise ValueError(f'{levels=} < 0')
+    if levels == 0:
+        return mean, real, imag
+    if samples < 1:
+        raise ValueError(f'{samples=} < 1')
+
+    if mean.shape != real.shape[-mean.ndim if mean.ndim else 1 :]:
+        raise ValueError(f'{mean.shape=} != {real.shape=}')
+    if real.shape != imag.shape:
+        raise ValueError(f'{real.shape=} != {imag.shape=}')
+
+    has_harmonic_axis = mean.ndim + 1 == real.ndim
+    if not has_harmonic_axis:
+        raise ValueError('No harmonic axis')
+    harmonic_max = real.shape[0]
+    if harmonic_max < 2:
+        raise ValueError(f'{harmonic=} < 2')
+    harmonics, _ = parse_harmonic(harmonic, harmonic_max)
+    harmonic_list = [
+        n for n in harmonics if 2 * n in range(1, harmonic_max + 1)
+    ]
```

I don't think this is quite correct. The `harmonics` input parameter describes which harmonics are contained in the `real` and `imag` arrays (which may be, for example, 1 and 2, or 2 and 4, or 1 to 4), but here it is assumed that they always include 1 to `harmonic_max`.

---

You are right @cgohlke! Thanks for pointing this out. How do you think the 'all' condition be handled? i am having some trouble thinking how to work with this. My understanding, also from the same parameter in other functions is that when `harmonic = 'all'`, it is expected to have the harmonics ordered from 1 to the max possible harmonic. Should this be more explicit in the description of the parameter? I think this concept is what 'lower harmonics' mean, but I think it would be good to say it more explicit here in the context of this function since it is very important the order of the harmonics for correct filtering.

---

Also, I think in the description of the function or in the harmonic parameter it should be detailed that for harmonics where `2 * harmonic` is not present, no filter is applied. But I am not sure how/where is the best way to describe this in the docstring.

---

>  for harmonics where 2 * harmonic is not present, no filter is applied

The filter requires h and 2h to work. It may be better to raise an error if that requirement is not met rather than silently not applying the filter.

---

Is explicitly allowing `harmonic = 'all'` really required in this function? Is the default `None` not enough? That is, by default, the phasor coordinates contain lower harmonics (`1...real.shape[0]`).

---

> The filter requires h and 2h to work. It may be better to raise an error if that requirement is not met rather than silently not applying the filter.

I agree, I don't like the idea of applying a filter only to certain harmonics and not others, so I will modify it to raise an error. This is also a good reason why searching for another implementation that doesn't need the `2 * harmonic` for all harmonics can be of great benefit here. Also making the function much simpler and broadly applicable.

> Is explicitly allowing `harmonic = 'all'` really required in this function? Is the default `None` not enough? That is, by default, the phasor coordinates contain lower harmonics (`1...real.shape[0]`).

You are right, now it doesn't make sense the `harmonic='all'` option. I will make the modifications.